src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
+
+            TokenType = JsonTokenType.None;
+            Value = ReadOnlySpan<byte>.Empty;
+            ValueType = JsonValueType.Unknown;
         }
 
-        public JsonReader(Utf8String str)
+        /// <summary>
+        /// Read the next token from the data buffer.
+        /// </summary>
+        /// <returns>True if the token was read successfully, else false.</returns>
+        public bool Read()
         {
-            _str = str.Trim();
-            _index = 0;
-            _insideObject = 0;
-            _insideArray = 0;
-            TokenType = 0;
-            _jsonStartIsObject = (byte)_str[0] == '{';
+            ref byte bytes = ref _buffer.DangerousGetPinnableReference();
+            int count = _buffer.Length;
+            int skip = SkipWhiteSpace(ref bytes, count);
+
+            ref byte next = ref Unsafe.Add(ref bytes, skip);
+            count -= skip;
+
+            int step = GetNextCharAscii(ref next, count, out char ch);
+            if (step == 0) return false;
+
+            switch (TokenType)
+            {
+                case JsonTokenType.None:
+                    if (ch == JsonConstants.OpenBrace)
+                        StartObject();
+                    else if (ch == JsonConstants.OpenBracket)
+                        StartArray();
+                    else
+                        throw new JsonReaderException();
+                    break;
+
+                case JsonTokenType.StartObject:
+                    if (ch == JsonConstants.CloseBrace)
+                        EndObject();
+                    else
+                        step = ConsumePropertyName(ref next, count);
+                    break;
+
+                case JsonTokenType.StartArray:
+                    if (ch == JsonConstants.CloseBracket)
+                        EndArray();
+                    else
+                        step = ConsumeValue(ch, step, ref next, count);
+                    break;
+
+                case JsonTokenType.PropertyName:
+                    step = ConsumeValue(ch, step, ref next, count);
+                    if (step == 0) return false;
+                    break;
+
+                case JsonTokenType.EndArray:
+                case JsonTokenType.EndObject:
+                case JsonTokenType.Value:
+                    step = ConsumeNext(ch, step, ref next, count);
+                    if (step == 0) return false;
+                    break;
+            }
+
+            _buffer = _buffer.Slice(skip + step);
+            return true;
         }
 
-        public JsonReader(string str)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void StartObject()
         {
-            _str = new Utf8String(str).Trim();
-            _index = 0;
-            _insideObject = 0;
-            _insideArray = 0;
-            TokenType = 0;
-            _jsonStartIsObject = (byte)_str[0] == '{';
+            if (_depth > MaxDepth)
+                throw new JsonReaderException();
```

Comment:
Yes. The JsonReaderException was actually a placeholder for something else. I just hadn't figured out what that was yet.